### PR TITLE
Add groups configuration synchronization status to `GET /agents/summary/status` API endpoint

### LINKED ISSUE #12363: Update `GET /agents/summary/status` response to include agent group configuration sync status
## Description

As part of #9334 , we need to update this endpoint response to include a summary of the agent group configuration sync status.

Currently, this endpoint returns the following response:

```json
{
  "data": {
    "active": 1,
    "disconnected": 0,
    "never_connected": 2,
    "pending": 0,
    "total": 3
  },
  "error": 0
}
```

The proposed new response would be the following:

```json
{
  "data":{
    "connection":{
      "active": 2,
      "disconnected": 0,
      "never_connected": 5,
      "pending": 0,
      "total": 7 
    },
    "configuration":{
      "synced": 2
      "not_synced": 5
      "total": 7
    }
  },
  "error": 0
}
```

## Requirements

The only requirement for this development is that this needs to be implemented in the absolute most efficient way, as this endpoint will be used a lot in the WUI. A full refactor may be considered.

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

### PR DESCRIPTION
**Pull request by mcarmona99**
|Related issue|
|---|
| #12363 |

This PR closes #12363.

In this pull request, the GET /agents/summary/status endpoint has been updated and now it returns the groups configuration synchronization status too.

Profiling has also been made and the theoretical limit of agents for which this endpoint would give a timeout has been tested (150.000 approx.).

`GET /agents/summary/status`

```
{
  "data": {
    "connection": {
      "active": 10,
      "disconnected": 0,
      "never_connected": 0,
      "pending": 0,
      "total": 10
    },
    "configuration": {
      "synced": 10,
      "total": 10,
      "not_synced": 0
    }
  },
  "error": 0
}
```

```
2022/03/03 09:19:00 INFO: wazuh 172.22.0.1 "GET /agents/summary/status" with parameters {} and body {} done in 0.016s: 200
```

### Test results

```python
$ pytest framework/wazuh/tests/test_agent.py  --disable-warnings 
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/manuel/git/wazuh/framework
plugins: html-3.1.1, metadata-1.11.0, cov-2.12.0, testinfra-5.0.0, asyncio-0.14.0
collected 117 items                                                                                                                                                              

framework/wazuh/tests/test_agent.py .....................................................................................................................                  [100%]

======================================================================== 117 passed, 9 warnings in 3.51s =========================================================================
```

```python
$ pytest framework/wazuh/core/tests/test_agent.py  --disable-warnings 
============================================================================== test session starts ===============================================================================
platform linux -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: /home/manuel/git/wazuh/framework
plugins: html-3.1.1, metadata-1.11.0, cov-2.12.0, testinfra-5.0.0, asyncio-0.14.0
collected 142 items                                                                                                                                                              

framework/wazuh/core/tests/test_agent.py ................................................................................................................................. [ 90%]
.............                                                                                                                                                              [100%]

======================================================================== 142 passed, 9 warnings in 0.92s =========================================================================
```

```
test_agent_GET_endpoints.tavern.yaml - Cluster environment 
	 92 passed, 93 warnings

test_rbac_black_agent_endpoints.tavern.yaml 
	 41 passed, 42 warnings

test_rbac_white_agent_endpoints.tavern.yaml 
	 41 passed, 42 warnings

test_rbac_white_all_endpoints.tavern.yaml 
	 161 passed, 162 warnings
```

### CODE REVIEW

File: framework/wazuh/tests/test_agent.py [hunk 1]
```diff
@@ -165,14 +165,19 @@ def test_agent_get_agents_summary_status(socket_mock, send_mock):
...
+
+    # For the following test cases, if summary_data has unexpected keys, a KeyError will be raised
+
+    # Check the data dictionary follows the expected keys schema
+    assert all(summary_data[key].keys() == expected_results[key].keys() for key in summary_data.keys()), \
+        'The result obtained has unexpected keys'
+    # Check that the agents count for connection and configuration statuses are the expected ones
+    assert all(all(summary_data[key][status] == expected_results[key][status] for status in
+                   summary_data[key].keys()) for key in summary_data.keys()), \
+        'The agents connection or configuration status counts are not the expected ones'
```
Comment (vicferpoy) [inline framework/wazuh/tests/test_agent.py:180]: This kind of `assert` is problematic without a previous condition. For instance, if we add a test mutant in the tested function:

```python
@expose_resources(actions=["agent:read"], resources=["agent:id:{agent_list}"], post_proc_func=None)
def get_agents_summary_status(agent_list: list = None):
    """Count the number of agents by connection and groups configuration synchronization statuses.

    Parameters
    ----------
    agent_list : list[str]
       List of agents ID's

    Returns
    -------
    WazuhResult
    """
    return WazuhResult({'data': {}})
```

The test will keep passing when this is not correct:

```shellsession
============================= test session starts ==============================
collecting ... collected 1 item

test_agent.py::test_agent_get_agents_summary_status PASSED               [100%]

======================== 1 passed, 10 warnings in 0.93s ========================
```

**Review by vicferpoy — CHANGES_REQUESTED** [review]

File: api/api/spec/spec.yaml [hunk 2]
```diff
@@ -7651,7 +7666,7 @@ paths:
         - $ref: '#/components/parameters/wait_for_complete'
       responses:
         '200':
-          description: "Get summary of agents"
+          description: "Get summary of agents connection and configuration statuses"
```
[the remark this answers is not in the document]
Comment (mcarmona99) [reply api/api/spec/spec.yaml:7682]: Done!

File: framework/wazuh/tests/test_agent.py (hunk 1 above)
Comment (mcarmona99) [reply framework/wazuh/tests/test_agent.py]: Changing the generator we are iterating will solve this issue.

Right now, I am iterating `summary_data.keys()`. It is here where, if `summary_data` is empty (case you commented (`{}`)), the test will wrongly pass.

If I use `expected_results.keys()` instead, the test will raise an `Exception` if `summary_data` doesn't have the same keys as `expected_results`. An `Exception` will also be raised if `summary_data` has more keys than `expected_results`.

Another solution could be adding `assert summary_data.keys() == expected_results.keys()` before these asserts.

Comment (vicferpoy) [reply framework/wazuh/tests/test_agent.py]: I always use the second solution with this kind of `assert`. Anyway, tests are best when you read them and know what they are doing in an instant. Using these structures with multiple `for` loops make that harder. Keep that in mind for further ones.

**Review by vicferpoy — APPROVED** [review]
LGTM

**Review by davidjiglesias — APPROVED** [review]

**Pull request merged**